packer/provisioner.go

```diff
@@ -55,6 +55,7 @@ func BasicPlaceholderData() map[string]string {
 	placeholderData := map[string]string{}
 	msg := "Build_%s. " + common.PlaceholderMsg
 	placeholderData["ID"] = fmt.Sprintf(msg, "ID")
+	placeholderData["SourceImageName"] = fmt.Sprintf(msg, "SourceImageName")
```

Comment:
OOOOh lesson time!  So, this is where we have default placeholder data _which is implemented in all of our builders_. But when I wrote this code I knew there would be a possibility that people wanted to add custom data that isn't available in all builders, so I wrote a secondary mechanism for adding information. Part of the interface changes we made in 1.5.0 was to make builder Prepare() methods return ([]string, []string, err) rather than ([]string, err). The first string is now reserved for a list of generated variables that the user is allowed to created. We use that list of variables to generate a custom placeholder map per builder that combines custom variables with the placeholder variables created in provisioner.go Check it out: 

https://github.com/hashicorp/packer/blob/master/packer/build.go#L153

This means that the right way to add SourceImageName to the amazon builders is to have it returned as the first return value from their Prepare() methods as `[]string{"SourceImageName"}` which will tell our validation to make room for it.  

But then we need to load it into the hookdata, and there's a mechanism for that, too: inside of the common step_provision, we load a generic "generated_data" value from the statebag: 
https://github.com/hashicorp/packer/blob/master/common/step_provision.go#L28-L33

This means that to make your custom SourceImageName available, we don't need to hardcode a check to state inside of this provision step (it should remain as generic as possible!). Instead, we can do something like `state.Put("generated_data", map[string]string{"SourceImageName": whateverTheVarIsNamedInTheBuilder })` inside of the amazon builder _before the call to step_provision_ and it will automatically be loaded into the hook data. 

If this doesn't make sense we can pair on it., but since we're only implementing this for one builder, I think it makes sense to use the custom fields rather than adding it to the common placeholders. The extra bonus of this is that if someone tries to use SourceImageName with a different builder than the AWS builders, validation will fail.

Reply:
Should we document this somewhere we could read it again easily?

Reply:
It might make sense to have basic instructions inside the "extending packer" docs

Reply:
Good, I do it in another PR